src/routing_table/mod.rs

```diff
@@ -632,17 +626,15 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
         };
         if bit_count == 0 || self.we_want_to_merge ||
            !self.sections.contains_key(&self.our_prefix.with_flipped_bit(bit_count - 1)) ||
-           (self.our_section.len() >= self.min_section_size &&
+           (!self.they_want_to_merge && self.our_section.len() >= self.min_section_size &&
             self.sections.iter().all(doesnt_need_to_merge_with_us)) {
             return None;
```

Comment:
This could be refactored into 2 separate `if` blocks each with their own `return None` line and a comment as to why: first block for the first three tests, and second block for the last three conditions chained with AND. (Or even a third block, for the "our prefix with last bit flipped doesn't exist.) More verbose, but hopefully more readable.